File: app/src/main/java/com/h3110w0r1d/t9launcher/model/AppViewModel.java
package com.h3110w0r1d.t9launcher.model;

import static com.h3110w0r1d.t9launcher.utils.ImageUtil.Drawable2IconBitmap;

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import androidx.preference.PreferenceManager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.h3110w0r1d.t9launcher.App;
import com.h3110w0r1d.t9launcher.R;
import com.h3110w0r1d.t9launcher.utils.DBHelper;
import com.h3110w0r1d.t9launcher.utils.Pinyin4jUtil;
import com.h3110w0r1d.t9launcher.vo.AppInfo;

import net.sourceforge.pinyin4j.format.HanyuPinyinCaseType;
import net.sourceforge.pinyin4j.format.HanyuPinyinToneType;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AppViewModel extends AndroidViewModel{

	private boolean isHideSystemApp;
	private final Set<String> hideAppList;

	public boolean isLoadingAppList = false;
	public boolean isShowingHideApp = false;

	private final SharedPreferences.Editor spEditor;
	private final Set<String> localPackageName = new HashSet<>();
	private final ArrayList<AppInfo> appList = new ArrayList<>();
	private final MutableLiveData<Boolean> Loading = new MutableLiveData<>(true);
	private final MutableLiveData<ArrayList<AppInfo>> searchResultLiveData = new MutableLiveData<>(new ArrayList<>());
	private final MutableLiveData<ArrayList<AppInfo>> hideAppListLiveData = new MutableLiveData<>(new ArrayList<>());
	public String SearchText = "";
	public SQLiteDatabase AppListDB;
	
	public AppViewModel(@NonNull Application application){
		super(application);
		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getApplication());
		spEditor = sharedPreferences.edit();
		hideAppList = sharedPreferences.getStringSet("hideAppList", new HashSet<>());
		isHideSystemApp =  sharedPreferences.getBoolean("hide_system_app", false);

		new Thread(()-> {
			Pinyin4jUtil.defaultFormat.setCaseType(HanyuPinyinCaseType.LOWERCASE);
			Pinyin4jUtil.defaultFormat.setToneType(HanyuPinyinToneType.WITHOUT_TONE);
			try {
				((App) getApplication()).appViewModel.AppListDB = new DBHelper((App) getApplication(), "AppList2.db", null, 1).getWritableDatabase();
			} catch (Exception e) {
				Toast.makeText(application, R.string.failed_init_database, Toast.LENGTH_LONG).show();
			}
			((App) getApplication()).appViewModel.loadAppList(getApplication());
		}).start();
	}

	public MutableLiveData<ArrayList<AppInfo>> getSearchResultLiveData(){
		return searchResultLiveData;
	}

	public MutableLiveData<ArrayList<AppInfo>> getHideAppListLiveData(){
		return hideAppListLiveData;
	}

	public void setAppHide(String packageName, boolean isHide){
		if (isHide) {
			hideAppList.add(packageName);
		} else {
			hideAppList.remove(packageName);
		}
	}

	public MutableLiveData<Boolean> getLoadingStatus(){
		return Loading;
	}

	public void setLoadingStatus(boolean loading){
		Loading.postValue(loading);
	}

	/**
	 * 加载并更新应用列表
	 */
	@SuppressLint("Recycle")
	public void loadAppList(Context context){
		if (isLoadingAppList) {
			return;
		}
		isLoadingAppList = true;
		appList.clear();
		Cursor cursor = AppListDB.query(
				"T_AppInfo",
				new String[]{
						"packageName", "appName", "startCount", "isSystemApp", "searchData"
				}, null, null, null, null, null);

		File data_dir = context.getDataDir();
		while (cursor.moveToNext()) {
			String packageName = cursor.getString(0);
			String appName = cursor.getString(1);
			int startCount = cursor.getInt(2);
			File app_icon = new File(data_dir, packageName + ".icon");
			if (!app_icon.exists()) {
				continue;
			}
			RoundedBitmapDrawable roundedDrawable = RoundedBitmapDrawableFactory.create(context.getResources(), app_icon.getAbsolutePath());
			roundedDrawable.getPaint().setAntiAlias(true);
			float cornerRadius = roundedDrawable.getIntrinsicHeight() * 0.26f;
			roundedDrawable.setCornerRadius(cornerRadius);

			boolean isSystemApp = cursor.getInt(3) == 1;
			Gson gson = new Gson();
			List<List<String>> searchData = gson.fromJson(cursor.getString(4),new TypeToken<List<List<String>>>(){}.getType());
			appList.add(new AppInfo(
					packageName, appName, startCount, roundedDrawable, isSystemApp, searchData
			));
		}
		cursor.close();
		if (!appList.isEmpty()) {
			setLoadingStatus(false);
		}
		RefreshAppInfo(context);
	}

	public void showDefaultAppList(){
		ArrayList<AppInfo> appInfo = new ArrayList<>();
		for (AppInfo app : appList) {
			if (isHideSystemApp && app.isSystemApp()){
				continue;
			}
			if (hideAppList.contains(app.getPackageName())) {
				continue;
			}
			appInfo.add(app);
		}
		appList.sort(new AppInfo.SortByStartCount());
		searchResultLiveData.postValue(appInfo);
	}
	
	/**
	 * 搜索应用
	 * @param key 关键词
	 */
	public void searchApp(@Nullable String key){
		if (key == null) {
			key = SearchText;
		} else {
			SearchText = key;
		}
		if (isShowingHideApp) {
			ShowHideApp();
			return;
		}
		if(TextUtils.isEmpty(key)){
			showDefaultAppList();
			return;
		}
		ArrayList<AppInfo> appInfo = new ArrayList<>();
		for(AppInfo app : appList){
			if (isHideSystemApp && app.isSystemApp()){
				continue;
			}
			if (hideAppList.contains(app.getPackageName())) {
				continue;
			}
			float matchRate = Pinyin4jUtil.Search(app, key); // 匹配度
			if (matchRate > 0) {
				app.setMatchRate(matchRate);
				appInfo.add(app);
			}
		}
		appInfo.sort(new AppInfo.SortByStartCount());
		appInfo.sort(new AppInfo.SortByMatchRate());

		searchResultLiveData.postValue(appInfo);
	}

	public void searchHideApp(String key){
		key = key.toLowerCase();
		ArrayList<AppInfo> appInfo = new ArrayList<>();
		for(AppInfo app : appList){
			if (hideAppList.contains(app.getPackageName()) && (app.getPackageName().contains(key) || app.getAppName().toLowerCase().contains(key))) {
				appInfo.add(app);
			}
		}
		for(AppInfo app : appList){
			if (!hideAppList.contains(app.getPackageName()) && (app.getPackageName().contains(key) || app.getAppName().toLowerCase().contains(key))) {
				appInfo.add(app);
			}
		}
		hideAppListLiveData.postValue(appInfo);
	}

	public boolean isAppHide(String packageName){
		return hideAppList.contains(packageName);
	}


	private void RefreshAppInfo(Context context){
		localPackageName.clear();
		PackageManager packageManager = context.getPackageManager();
		List<ResolveInfo> resolveInfoList = packageManager.queryIntentActivities(new Intent(Intent.ACTION_MAIN, null).addCategory(Intent.CATEGORY_LAUNCHER), 0);
		for (ResolveInfo resolveInfo : resolveInfoList) {
			localPackageName.add(resolveInfo.activityInfo.packageName);
		}

		Cursor cursor = AppListDB.query(
				"T_AppInfo",
				new String[]{
						"packageName", "appName", "startCount", "isSystemApp", "searchData"
				}, null, null, null, null, null);
		List<String> needRemove = new ArrayList<>();
		while (cursor.moveToNext()) {
			String packageName = cursor.getString(0);
			if (!localPackageName.contains(packageName)) {
				needRemove.add(packageName);
			}
		}
		cursor.close();
		for (int i=appList.size()-1; i>=0; i--){
			AppInfo app = appList.get(i);
			if (needRemove.contains(app.getPackageName())){
				AppListDB.delete("T_AppInfo", "packageName = ?", new String[]{app.getPackageName()});
				appList.remove(app);
			}
		}
		setLoadingStatus(false);

		searchApp(null);

		new Thread(() -> {
			File data_dir = context.getDataDir();
			for(ResolveInfo resolveInfo : resolveInfoList){
				try{
					String packageName = resolveInfo.activityInfo.packageName;
					PackageInfo packageInfo = packageManager.getPackageInfo(packageName, 0);
					String appName = resolveInfo.loadLabel(packageManager).toString();
					Drawable icon = resolveInfo.loadIcon(packageManager);
					RoundedBitmapDrawable roundedDrawable = RoundedBitmapDrawableFactory.create(context.getResources(), Drawable2IconBitmap(icon));
					roundedDrawable.getPaint().setAntiAlias(true);
					float cornerRadius = roundedDrawable.getIntrinsicHeight() * 0.26f;
					roundedDrawable.setCornerRadius(cornerRadius);

					File app_icon = new File(data_dir, packageName + ".icon");
					try (FileOutputStream out = new FileOutputStream(app_icon)) {
						assert roundedDrawable.getBitmap() != null;
						roundedDrawable.getBitmap().compress(Bitmap.CompressFormat.PNG, 100, out);
					} catch (IOException e) {
						e.printStackTrace();
					}
					List<List<String>> searchData = Pinyin4jUtil.getPinYin(appName);
					ApplicationInfo applicationInfo = packageInfo.applicationInfo;
					boolean isSystemApp = (applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0;
					boolean findInAppList = false;
					for (AppInfo app : appList){
						if (app.getPackageName().equals(packageName)){
							findInAppList = true;
							app.setAppName(appName);
							app.setAppIcon(roundedDrawable);
							app.setIsSystemApp(isSystemApp);
							app.setSearchData(searchData);
							InsertOrUpdate(app);
							break;
						}
					}
					if (findInAppList){
						continue;
					}
					AppInfo app = new AppInfo(packageName, appName, 0, roundedDrawable, isSystemApp, searchData);
					appList.add(app);
					InsertOrUpdate(app);
				}catch(PackageManager.NameNotFoundException e){
					e.printStackTrace();
				}
			}

			searchApp(null);

			isLoadingAppList = false;
		}).start();

	}

	public void InsertOrUpdate(AppInfo app) {
		String packageName = app.getPackageName();
		String appName = app.getAppName();
		Bitmap appIcon = ((RoundedBitmapDrawable)app.getAppIcon()).getBitmap();
		boolean isSystemApp = app.isSystemApp();
		List<List<String>> searchData = app.getSearchData();
		try {
			assert appIcon != null;
			AppListDB.execSQL("INSERT INTO T_AppInfo VALUES(?,?,?,?,?)",
					new Object[]{
							packageName, appName, 0, isSystemApp ? 1 : 0, new Gson().toJson(searchData)
					});
		} catch (Exception e) {
			AppListDB.execSQL("UPDATE T_AppInfo SET appName = ?, isSystemApp = ?, searchData = ? WHERE packageName = ?",
					new Object[]{
							appName, isSystemApp ? 1 : 0, new Gson().toJson(searchData), packageName
					});
		}
	}

	public void UpdateStartCount(AppInfo app) {
		String packageName = app.getPackageName();
		int startCount = app.getStartCount();
		try {
			AppListDB.execSQL("UPDATE T_AppInfo SET startCount = ? WHERE packageName = ?",
					new Object[]{
							startCount, packageName
					});
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public void SaveHideList() {
		spEditor.remove("hideAppList").commit();
		spEditor.putStringSet("hideAppList", hideAppList).commit();
	}

	public void ShowHideApp() {
		isShowingHideApp = true;
		ArrayList<AppInfo> appInfo = new ArrayList<>();
		for (AppInfo app : appList){
			if (hideAppList.contains(app.getPackageName())){
				appInfo.add(app);
			}
		}
		if (isHideSystemApp){
			for (AppInfo app : appList){
				if (!hideAppList.contains(app.getPackageName()) && app.isSystemApp()){
					appInfo.add(app);
				}
			}
		}
		searchResultLiveData.postValue(appInfo);
	}

	public void setHideSystemApp(boolean newValue) {
		isHideSystemApp = newValue;
	}

}

File: app/src/main/java/com/h3110w0r1d/t9launcher/model/AppViewModel.kt
package com.h3110w0r1d.t9launcher.model

import androidx.compose.runtime.staticCompositionLocalOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.h3110w0r1d.t9launcher.data.app.AppInfo
import com.h3110w0r1d.t9launcher.data.app.AppInfo.SortByMatchRate
import com.h3110w0r1d.t9launcher.data.app.AppRepository
import com.h3110w0r1d.t9launcher.data.config.AppConfig
import com.h3110w0r1d.t9launcher.data.config.AppConfigManager
import com.h3110w0r1d.t9launcher.data.config.AppListStyleConfig
import com.h3110w0r1d.t9launcher.data.config.KeyboardStyleConfig
import com.h3110w0r1d.t9launcher.data.config.SearchConfig
import com.h3110w0r1d.t9launcher.data.config.ThemeConfig
import com.h3110w0r1d.t9launcher.utils.PinyinUtil
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.Locale
import javax.inject.Inject

@HiltViewModel
class AppViewModel
    @Inject
    constructor(
        private val appRepository: AppRepository,
        private val configManager: AppConfigManager,
        private val pinyinUtil: PinyinUtil,
    ) : ViewModel() {
        private var searchText: String = ""
        private var isLoadingAppList: Boolean = false
        private var isAppListListenerStarted: Boolean = false

        val appConfig: StateFlow<AppConfig> = configManager.appConfig

        val appMap: StateFlow<HashMap<String, AppInfo>> = appRepository.appMap
        private val _isLoading: MutableStateFlow<Boolean> = MutableStateFlow(true)
        val isLoading: StateFlow<Boolean> = _isLoading
        private val _isRefreshing: MutableStateFlow<Boolean> = MutableStateFlow(false)
        val isRefreshing: StateFlow<Boolean> = _isRefreshing
        private val _searchResultAppList: MutableStateFlow<ArrayList<AppInfo>> = MutableStateFlow(ArrayList())
        val searchResultAppList: StateFlow<ArrayList<AppInfo>> = _searchResultAppList

        private val _hideAppList: MutableStateFlow<ArrayList<AppInfo>> =
            MutableStateFlow(ArrayList())
        val hideAppList: StateFlow<ArrayList<AppInfo>> = _hideAppList

        fun switchAppHide(app: AppInfo) {
            val config = appConfig.value
            if (isAppHide(app)) {
                val newHiddenIds =
                    config.search.hiddenComponentIds.toMutableSet().apply {
                        remove(app.componentId())
                    }
                viewModelScope.launch {
                    configManager.updateSearchConfig(
                        config.search.copy(hiddenComponentIds = newHiddenIds),
                    )
                }
            } else {
                val newHiddenIds =
                    config.search.hiddenComponentIds.toMutableSet().apply {
                        add(app.componentId())
                    }
                viewModelScope.launch {
                    configManager.updateSearchConfig(
                        config.search.copy(hiddenComponentIds = newHiddenIds),
                    )
                }
            }
        }

        fun setLoadingStatus(loading: Boolean) {
            _isLoading.value = loading
        }

        /**
         * 启动应用列表监听器，当应用列表更新时自动刷新搜索结果
         */
        private fun startAppListListener() {
            if (isAppListListenerStarted) return

            isAppListListenerStarted = true
            viewModelScope.launch {
                appRepository.appList.collect { appList ->
                    if (appList.isNotEmpty()) {
                        setLoadingStatus(false)
                        searchApp(null)
                    }
                }
            }
        }

        // 简化的配置更新方法
        fun updateAppListStyle(config: AppListStyleConfig) {
            viewModelScope.launch {
                configManager.updateAppListStyle(config)
            }
        }

        fun updateKeyboardStyle(config: KeyboardStyleConfig) {
            viewModelScope.launch {
                configManager.updateKeyboardStyle(config)
            }
        }

        fun updateThemeConfig(config: ThemeConfig) {
            viewModelScope.launch {
                configManager.updateThemeConfig(config)
            }
        }

        fun updateSearchConfig(config: SearchConfig) {
            viewModelScope.launch {
                configManager.updateSearchConfig(config)
            }
        }

        fun setQuickStartApp(
            index: Int,
            componentId: String,
        ) {
            val config = appConfig.value
            val newQuickStartConfig = config.shortcutConfig.toMutableList()
            newQuickStartConfig[index] = componentId
            viewModelScope.launch {
                configManager.updateShortcutConfig(ArrayList(newQuickStartConfig))
            }
        }

        /**
         * 加载并更新应用列表
         */
        fun loadAppList() {
            if (isLoadingAppList || !isLoading.value) {
                return
            }

            // 启动应用列表监听器
            startAppListListener()

            viewModelScope.launch {
                isLoadingAppList = true
                appRepository.loadAllApps()

                // 检查是否有数据
                val currentAppList = appRepository.appList.value
                if (currentAppList.isNotEmpty()) {
                    setLoadingStatus(false)
                    searchApp(null)
                }

                appRepository.updateAppInfo(false)
                setLoadingStatus(false)
                searchApp(null)
                isLoadingAppList = false
            }
        }

        fun refresh() {
            viewModelScope.launch {
                _isRefreshing.value = true
                withContext(Dispatchers.IO) {
                    appRepository.updateAppInfo(true)
                }
                searchApp(null)
                _isRefreshing.value = false
            }
        }

        fun showDefaultAppList() {
            val appInfo = ArrayList<AppInfo>()
            val currentAppList = appRepository.appList.value
            val config = appConfig.value

            for (app in currentAppList) {
                if (config.search.hideSystemAppEnabled && app.isSystemApp) {
                    continue
                }
                if (isAppHide(app)) {
                    continue
                }
                app.setMatchRange(-1, -1)
                appInfo.add(app)
            }
            _searchResultAppList.value = appInfo
        }

        fun searchApp(key: String?): Boolean {
            var key = key
            if (key == null) {
                key = searchText
            }
            if (key.isEmpty()) {
                showDefaultAppList()
                searchText = key
                return true
            }
            val appInfo = ArrayList<AppInfo>()
            val currentAppList = appRepository.appList.value
            val config = appConfig.value

            for (app in currentAppList) {
                if (config.search.hideSystemAppEnabled && app.isSystemApp) {
                    continue
                }
                if (config.search.hiddenComponentIds.contains(app.componentId())) {
                    continue
                }
                val matchRate = pinyinUtil.search(app, key, config.search.englishFuzzyMatchEnabled) // 匹配度
                if (matchRate > 0) {
                    app.matchRate = matchRate
                    appInfo.add(app)
                }
            }
            if (appInfo.isEmpty()) {
                return false
            }
            appInfo.sortWith(SortByMatchRate())
            _searchResultAppList.value = appInfo
            searchText = key
            return true
        }

        fun setShowedOnboarding() {
            viewModelScope.launch {
                configManager.setShowedOnboarding()
            }
        }

        fun searchHideApp(key: String) {
            var key = key
            key = key.lowercase(Locale.getDefault())
            val appInfo = ArrayList<AppInfo>()
            val currentAppList = appRepository.appList.value
            val config = appConfig.value

            for (app in currentAppList) {
                if (config.search.hideSystemAppEnabled && app.isSystemApp) {
                    continue
                }
                if (app.packageName
                        .contains(key) ||
                    app.appName
                        .lowercase(
                            Locale.getDefault(),
                        ).contains(key)
                ) {
                    appInfo.add(app)
                }
            }
            appInfo.sortWith { o1, o2 ->
                (if (isAppHide(o2)) 1 else 0) - (if (isAppHide(o1)) 1 else 0)
            }
            _hideAppList.value = appInfo
        }

        private fun isAppHide(app: AppInfo): Boolean {
            val config = appConfig.value
            return config.search.hiddenComponentIds.contains(app.componentId())
        }

        fun updateStartCount(app: AppInfo) {
            app.startCount += 1
            appRepository.updateStartCount(app)
        }

        fun showHideApp() {
            val appInfo = ArrayList<AppInfo>()
            val currentAppList = appRepository.appList.value
            val config = appConfig.value

            for (app in currentAppList) {
                if (isAppHide(app) || (config.search.hideSystemAppEnabled && app.isSystemApp)) {
                    appInfo.add(app)
                    app.setMatchRange(-1, -1)
                }
            }
            _searchResultAppList.value = appInfo
        }
    }

val LocalGlobalViewModel =
    staticCompositionLocalOf<AppViewModel> {
        error("AppViewModel not provided!")
    }
